### DISCUSSION TITLE
Struct tags implementation

### DISCUSSION THREAD

**Discussion by Tracnac** [sijms/go-ora#132] [Ideas]
Hello,

First thanks for this great project.

A little question do you have any plan to implement values to struct tagging in your project.
From this great idea : [github.com/blockloop/scan](https://github.com/blockloop/scan)
And especially the **RowStrict** function...

Example:

```go
var name struct {
	Name string `db: "NAME"`
}

// Connection...
        stmt := go_ora.NewStmt(`select 'go-ora' as NAME from dual`, conn)
	defer stmt.Close()

	rows, err := stmt.Query(nil)
        // err
	defer rows.Close()

        // Automatic matching from value to the tag
        go_ora.RowStrict(&name, rows)
```

Thanks Yvan

**Comment by sijms** [comment]
I actually made something like this in User Defined Type see example UDT
but no problem I will add it for query also

**Reply by Tracnac** [reply]
Thank you, I did not notice this example, I will take a look.
But I agree much simple if Scan is implemented at the Query => rows level.

Thanks Yvan

**Comment by sijms** [comment]
I make a commit to add support for struct with tag in scan function
```golang
type test struct{
    Number int64 `db:"name:num"`
    Name   string `db:"name:tname"`
}

func main() {
    connStr := "oracle://user:pass@server/service_name"
    conn, err = go_ora.NewConnection(connStr)
    if err != nil {
        fmt.Println("Can't create connection: ", err)
	return
    }
    err = conn.Open()
    if err != nil {
        fmt.Println("Can't open connection: ", err)
        return
    }
    defer func() {
        err = conn.Close()
        if err != nil {
            fmt.Println("Can't close connection: ", err)
        }
    }()
    stmt := go_ora.NewStmt("select rownum as num, 'test' as tname from all_objects where rownum <= 140", DB)
    defer func(){
        err = stmt.Close()
        if err != nil {
            fmt.Println("Can't create stmt: ", err)
        }
    }()
    rows, err := stmt.Query_(nil)
    if err != nil {
        fmt.Println("Can't query: ", err)
        return
    }
    for rows.Next_() { 
        err = rows.Scan(&t)
        if err != nil {
	        fmt.Println(err)
	        return
        }
        fmt.Println(t)
    }
}

**Comment by sijms** [comment]
I add an example  query_to_struct

**Comment by Tracnac** [comment]
Thank you for your update.
I will check, and use it of course :)

Yvan

**Comment by Tracnac** [comment]
Hi,

Not sure what happend here... Could you help me.
The first query works perfectly filling the tag, but for the second query, I expect to have s2c1 and s2c2 to be set by the result.

Thank you
Yvan

As always a simplified code to explain what happend:

```go
package main

import (
	"fmt"
	"os"

	go_ora "github.com/sijms/go-ora/v2"
)

type db_TS struct {
	S1C1 string `db:"name:s1c1"`
	S1C2 string `db:"name:s1c2"`
	S2C1 string `db:"name:s2c1"`
	S2C2 string `db:"name:s2c2"`
}

func main() {
	var (
		err     error
		stmt    *go_ora.Stmt
		queries []string
		db      db_TS
	)

	queries = append(queries,
		"select 's1c1' as s1c1, 's1c2' as s1c2 from dual",
		"select 's2c1' as s2c1, 's2c2' as s2c2 from dual")

	MsgErrorDie := func(msgText string, msgError error) {
		if msgError != nil {
			fmt.Println(msgText, msgError)
			os.Exit(1)
		}
	}

	DB, err := go_ora.NewConnection("oracle://tracnac:tracnac@192.168.122.90/xepdb1")
	MsgErrorDie("URL error: ", err)

	err = DB.Open()
	MsgErrorDie("DB open error: ", err)
	defer DB.Close()

	for _, query := range queries {
		stmt = go_ora.NewStmt(query, DB)
		defer stmt.Close()

		rows, err := stmt.Query_(nil)
		MsgErrorDie("Query error: ", err)
		defer rows.Close()

		rowsCount := 0
		for rows.Next_() {
			rows.Scan(&db)
			rowsCount += 1
		}
		fmt.Println("Rows count: ", rowsCount)
	}
	fmt.Printf("%#v", db)

// Output: main.db_TS{S1C1:"s1c1", S1C2:"s1c2", S2C1:"", S2C2:""} 
// Instead of main.db_TS{S1C1:"s1c1", S1C2:"s1c2", S2C1:"s2c1", S2C2:"s2c2"}
}

```

**Reply by Tracnac** [reply]
Forget go.mod, don't know if it's important...

```
module test

require github.com/sijms/go-ora/v2 v2.2.16-0.20211219220758-025150f9dca6

go 1.16
```

**Reply by sijms** [reply]
module need to make new release I will make after finish testing and fixing

**Comment by Tracnac** [comment]
Humm, sorry completely forgot to catch the rows.Scan error.
```
Query error:  go-ora: mismatching between Scan function input count and column count
```
Understand now what's happening, is it possible to just count the number of field that match the select with the tag, or even better just ignore all tags that not matching at all...
Not sure if I'm clear...

For example:
```go
struct {
 id int
 anotherField string
 ...
 CustomerName string `db:"name:c_name`
 CustomerAdress string `db:"name:c_adress`
  ....
  ProductID int `db:"name:p_id`
  ProductName string `db:"name:p_name`
...
etc...
}
```
with queries:

select name as c_name, adress as c_adress from customers
select id as p_id, name as p_name from products
etc...

Just taking the tag that match colunm and ignore everything else

Yvan

**Comment by sijms** [comment]
I make a commit that bypass mismatched field name and also bypass extra tags (example: columns are 2 and tags are 4). the last one is the cause of the error above. test and if ok I will make a new release

**Comment by Tracnac** [comment]
Super, thank you
I'll check some other use case.
Keep you in touch

Thank you very much
Yvan

```
main.db_TS{S1C1:"s1c1", S1C2:"s1c2", S2C1:"s2c1", S2C2:"s2c2"}
```
with:
```
require github.com/sijms/go-ora/v2 v2.2.16-0.20211220161318-c50eaeb66035
```

**Comment by Tracnac** [comment]
Do you plan to support nested struct ?

Like:
```go
type db_TS struct {
	S1C1     string `db:"name:s1c1"`
	S1C2     string `db:"name:s1c2"`
	AnotherStruct struct {
		S2C1 string `db:"name:s2c1"`
		S2C2 string `db:"name:s2c2"`
	}
}
```

Thanks
Yvan

**Reply by sijms** [reply]
I think better to keep it simple
and also it is support by the way you can pass first arg db_TS object then second arg db_TS.AnotherStruct object

**Reply by Tracnac** [reply]
Yes I understand but sometime it's not so easy like:

```
type db_TS struct {
	S1C1     string `db:"name:s1c1"`
	S1C2     string `db:"name:s1c2"`
	S2C1 string `db:"name:s2c1"`
	AnotherStruct struct {
		S2C2 string `db:"name:s2c2"`
	}
}
```
Thank you for your time.

**Comment by Tracnac** [comment]
Don't know if it's related or I missed something but I have:
`Query error:  go-ora: column 0 require an integer`

It's not possible to pass a var directly ?

With:

```go
	var DBID int = 0
	stmt := go_ora.NewStmt("select 1 as DBID from dual", DB)   // It's an integer, an Oracle number to be more precise.
	defer stmt.Close()

	rows, err := stmt.Query_(nil)
	MsgErrorDie("Query error: ", err)
	defer rows.Close()

	err = rows.Scan(&DBID)
	MsgErrorDie("Query error: ", err)
	fmt.Printf("%#v", DBID)
```

Yvan

**Reply by sijms** [reply]
you need to call `rows.Next_()` at least one time before  call `rows.Scan(&DBID)`

**Reply by Tracnac** [reply]
Ouch sorry... I missed this one.

I'm tired fixing log4j on hundreds of servers :)

Thanks Yvan

**Comment by sijms** [comment]
I make a new release 2.2.16

**Comment by Tracnac** [comment]
> I make a new release 2.2.16

Thanks Samy for your time and availability.

Yvan

**Reply by sijms** [reply]
you are welcome
